pkg/autoupgrade/tags.go

```diff
@@ -41,8 +41,27 @@ func findLatestTagForImageWithPattern(ctx context.Context, c daemonClient, names
 		return "", false, err
 	}
 
-	newTag, err := FindLatest(imagename.DefaultTag, pattern, tags)
-	return strings.TrimPrefix(ref.Context().Tag(newTag).Name(), defaultNoReg+"/"), newTag != imagename.DefaultTag, err
+	newTag := imagename.DefaultTag
+	newImage := strings.TrimPrefix(ref.Context().Tag(newTag).Name(), defaultNoReg+"/")
+	for len(tags) > 0 {
+		nTag, err := FindLatest(imagename.DefaultTag, pattern, tags)
+		if err != nil || nTag == imagename.DefaultTag {
+			// resorting to default tag, so stop trying (we don't want to loop forever)
+			break
+		}
+		img := strings.TrimPrefix(ref.Context().Tag(nTag).Name(), defaultNoReg+"/")
+		if err := c.checkImageAllowed(ctx, namespace, img); err != nil {
+			// remove the tag from the list and try again
+			tags = slices.Filter(nil, tags, func(tag string) bool { return tag != nTag })
+		} else {
+			// found a valid tag that is allowed by all rules, so use it
+			newTag = nTag
+			newImage = img
+			break
+		}
+	}
```

Comment:
Can this logic be incorporated into `FindLatest` so that the tag matching happens as few times a possible?

Reply:
Unfortunately I couldn't find any cleaner or more performant way of doing this.
Copied two options from Slack:

>a) Check for every tag -> that's not really performant, but at least linear
b) Check only for latest tag -> In that case, we'd need to keep a sorted list, to cycle through it in reverse or we will end up with the same issue
I see that (b) can achieve way better performance in all non-worst cases, but would also force us to go from "keep note of the latest tag" to "keep a sorted list" in the current logic.

We aligned on keeping it as it is for now.